### ISSUE TITLE
Feature request:  Enable Minikeepass as a Safari extension thru actions menu

### ISSUE THREAD

**Issue by loneyliver** [MiniKeePass/MiniKeePass#335]
I experimented with 1Password for IOS8 and they have enabled it as a safari extension  which allows you to fill in passwords for safari via the actions menu. It works quite well. Was hoping this kind of feature can be integrated into Minikeepass

**Comment by azetilov** [comment]
Ok, so I started work on this feature and here's quick summary of my progress:
* Javascript side - can find and fill forms on websites
* JS <--> Extension communication
* Can access user databases (stored in application) 
* UI draft

Here're a couple of screenshots to give you an idea of how it looks:
![safariextension basicflow](https://cloud.githubusercontent.com/assets/3044953/5740137/55bd4260-9c09-11e4-9663-6abf622884be.png)

Right now I'm working on to integrate extension with the application infrastructure (settings support, re-using common UI bits, etc).
As discussed with @jarush, there will be points to refactor and DRY out some common code.

One of the things that needs a better thought is the UX of extension:
* I don't really like the idea of making user to choose a database every time he/she wants to get a password for a website
* Due to specifics of iOS App Extensions (they have very limited lifecycle) extension is not able to hold database open and now user is required to unlock database each time - I think it's not acceptable and needs to be resolved

I'm not an expert here, but to my view, ideal workflow would be:
**First launch**:
1. Being on the website with a form in Safari, user hits MiniKeePass extension icon
2. Extension presents a database to unlock (and remembers it, somehow)
3. User unlocks database
4. Extension looks up credentials for the given website and fills a website form

**Subsequent launches:**
1. Being on the website with a form in Safari, user hits MiniKeePass extension icon
2. Extension uses opened database to lookup credentials and fill a website form

Would appreciate if someone could give an advice here.

What do you think, guys?
Do you have any ideas on this?

**Comment by wooster11** [comment]
Selecting a database every time does seem a bit cumbersome. But if that's a limitation of iOS, then there's not much you can do. What could help is to add support for Touch ID. Then at least the password doesn't need to be inputted every time.

**Comment by azetilov** [comment]
Actually, selecting database each time is not a limitation of iOS, it's just the way it is currently implemented in MiniKeePass. The limitation is that I cannot keep database open whole time. Having Touch ID or PIN-lock enabled would solve the problem for those who allowed MiniKeePass to remember passwords for databases. But that is not often the case, as it is considered less secure to store you master password anywhere else than in your head. I will look for ways to work around this

As for database selection, my assumption here is that most of the users have only one database with passwords that they use frequently. That's why I thought it would be good to have some option to omit *select database* screen and go straight to *unlock specific database* screen.
I'm not sure, however, if we should really care a lot about this at this stage. At least for the first version.

BTW, I've seen some fingerprints of Touch ID in the latest source code, so it's likely that it will be added at some point soon.

**Comment by loneyliver** [comment]
First off your work on this is appreciated. I agree that Touch ID would be an integral part of the whole process. As azetilov mentioned, most have one database, it would be better to have immediate access to that database or if there are more than one, then be able to assign a primary database to default to.

**Comment by nazriel** [comment]
@loneyliver ping. What's the status of this?

**Comment by nomadhacker** [comment]
@azetilov this will be great when complete. status?

**Comment by gen251426** [comment]
This would be a killer feature. Any updates?

**Comment by azetilov** [comment]
I was a bit busy lately, but hoping to renew work on this till EOM.

As for the progress, you can look here:
https://github.com/azetilov/MiniKeePass/tree/WebExtension 
What's there already:
1. JavaScript part
2. File sharing mechanism between App and Extension
3. Database selection view
4. Icons and other minor stuff

What is especially good is that recently Touch ID support was released and WebExtension can rely on this feature to unlock databases.

**Comment by gen251426** [comment]
Many many thanks and all your hard work is much appreciated. 






On Tue, Nov 3, 2015 at 3:34 AM -0800, "azetilov" <notifications@github.com> wrote:












I was a bit busy lately, but hoping to renew work on this till EOM.



As for the progress, you can look here:

https://github.com/azetilov/MiniKeePass/tree/WebExtension 

What's there already:

1. JavaScript part

2. File sharing mechanism between App and Extension

3. Database selection view

4. Icons and other minor stuff



What is especially good is that recently Touch ID support was released and WebExtension can rely on this feature to unlock databases.

**Comment by Designator-Ol** [comment]
@azetilov Many thanks for your hard work on this. I would love to switch completely to KeePass and MiniKeePass from 1Password but the currently missing webextension is holding me back. Can you provide is an eta for when this will be released ? 
Again thanks for you work :+1:

**Comment by azetilov** [comment]
**Status update**:
 - Implemented 'Select database' screen (with a shortcut for single database scenario)
 - Implemented 'Unlock database' screen (password or key-file)
 - Implemented three-way search (full URL/domain/naked domain)

**To-do**:
 - New code needs refactoring + extract common parts with the existing app
 - Integrate form-filling on Safari side
 - Implement 'Select entry' screen - see below
 
**High-level feature outlook**:
 - Add support for 1Password's extension (see [here](https://blog.agilebits.com/2014/08/04/filling-with-your-approval-on-1passwords-app-extension-and-ios-8-security) - specifically, chapter 'We’re using a brand-neutral URL scheme on purpose')
 - Button to open MiniKeePass

**One topic to mention**:
When search returns multiple entries for one url, extension will present 'Select entry' screen that will give user ability to choose credentials to use in this specific case. It's OK for multiple entries, but when single entry was found, it seems natural not to present a selection screen and simply return credentials to Safari and fill out the form.
However, it's only secure if extension is going to be used in Safari, not other apps. If we ever integrate the 1Password's extension protocol, any app could possibly enumerate user's database and get passwords in a sort of hidden way. 
More context on this and why 1Password guys chose to always present 'Select entry' screen:
[Filling with your approval: On 1Password’s App Extension and iOS 8 security](https://blog.agilebits.com/2014/08/04/filling-with-your-approval-on-1passwords-app-extension-and-ios-8-security/)

**Comment by eternia16** [comment]
Just out of curiosity, how far are you on your feature?

**Comment by liuwren** [comment]
I can't wait to try the new MiniKeePass with Safari extension! Actually, it is something like URL Scheme, is it ?

**Comment by Thumpxr** [comment]
Any update on this?

**Comment by chrisTRD** [comment]
This feature is a must-have on iOS and I think the lack of any form of browser extension is really hurting the iOS implementation of KeePass right now.

**Comment by Thumpxr** [comment]
Yes, but it seems like the developers are inactive or not interested.

Am Donnerstag, 3. November 2016 schrieb chrisTRD :

**Comment by stevenleeg** [comment]
I'd just like to add my +1 for this feature. Having a safari extension would be an amazing way to sell the KeePass ecosystem as a feature-complete alternative to 1Password.

**Comment by jaege** [comment]
Any update in 2017?

**Comment by ericmoret** [comment]
Could you please release this extension

**Comment by rqchew** [comment]
Another voice in this

**Comment by mbrockman1** [comment]
I hate to be a pain but I am in need of this functionality. I am willing to help if given a status update!

**Comment by chaosbunker** [comment]
Any updates on this?

**Comment by metbril** [comment]
I think this is the one feature that's holding me back from stopping thinking about 1Password and LastPass as alternatives. I take manually copying and pasting the database for granted but from a usability point of view this one feature is holding me back from jumping all-in to the KeePass ecosystem.

**Comment by Mettbrot** [comment]
@azetilov is there any news on this? Is there currently an option for minikeepass to insert passwords without needing to copy them?
